clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.filesystem.hadoop;
+
+import com.datastrato.gravitino.Catalog;
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.client.FilesetCatalog;
+import com.datastrato.gravitino.client.GravitinoClient;
+import com.datastrato.gravitino.client.GravitinoMetaLake;
+import com.datastrato.gravitino.file.Fileset;
+import com.datastrato.gravitino.shaded.com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.InvalidPathException;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.util.Progressable;
+
+public class GravitinoVirtualFileSystem extends FileSystem {
+  private FileSystem proxyFileSystem;
+  private Path workingDirectory;
+  private URI uri;
+  private NameIdentifier filesetIdentifier;
+
+  @Override
+  public void initialize(URI name, Configuration conf) throws IOException {
+    if (name.toString().startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      Preconditions.checkArgument(
+          conf.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_SERVER_URI_KEY) != null,
+          "Gravitino server uri is not set in the configuration");
+      this.filesetIdentifier = normalizedIdentifier(name.getPath());
+      String serverUri =
+          conf.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_SERVER_URI_KEY);
+      // TODO Need support more authentication types, now we only support simple auth
+      try (GravitinoClient client = GravitinoClient.builder(serverUri).withSimpleAuth().build()) {
+        GravitinoMetaLake metaLake =
+            client.loadMetalake(NameIdentifier.of(filesetIdentifier.namespace().level(0)));
+        Catalog catalog =
+            metaLake.loadCatalog(
+                NameIdentifier.of(
+                    filesetIdentifier.namespace().level(0),
+                    filesetIdentifier.namespace().level(1)));
+        FilesetCatalog filesetCatalog = (FilesetCatalog) catalog.asFilesetCatalog();
+        Fileset fileset = filesetCatalog.loadFileset(filesetIdentifier);
+        String storageLocation = fileset.storageLocation();
+        Preconditions.checkArgument(
+            storageLocation != null
+                && storageLocation.startsWith(
+                    GravitinoVirtualFileSystemConfiguration.HDFS_SCHEME_PREFIX),
```

Comment:
Do we only support HDFS?

Reply:
In the first version, I recommend supporting HDFS first, and in the future, other storages that implement HCFS can be supported.

Reply:
Maybe we should create an issue to track it.

Reply:
#2310 Done